### ISSUE TITLE
Fitting fails when 'min=' attribute is included for one parameter

### ISSUE THREAD

**Issue by CarpeCimex** [lmfit/lmfit-py#278]
I'm really impressed with lmfit and I'm incorporating it into a general-purpose model fitter for electronic networks.  Thank you for a great addition to the Python pantheon.

There seems to be an odd bug when I use the 'min=' attribute on one particular parameter, though.  Here is a test script (please rename to .py):
[lmfit_bug.txt](https://github.com/lmfit/lmfit-py/files/40908/lmfit_bug.txt)
and here is the data file to accompany it (please rename to .csv):
[Sample.txt](https://github.com/lmfit/lmfit-py/files/40909/Sample.txt)

Lines 28 and 29 either cause the bug (when min= is used) or don't (when min= is absent).  Comment one or the other line out.  This is only a problem with the 'Cp' parameter.  I have tried min with large negative numbers, and it still causes fitting to fail until I get to -np.inf.

(I know the "should be" values are correct by plugging them into the model, plotting, and seeing it line up with the data.)

Please let me know if there is a workaround for this, or if I'm doing something wrong.  

Thanks very much,
Gerrit

**Comment by newville** [comment]
@CarpeCimex  this is basically due to floating point precision and the use of very small numbers.   That is, the easiest solution would be to have your Cp be closer to 1, perhaps changing your objective function to have:
```python
 Ycp = 1j*w*1.e12*Cp + 1/Rp
```
(even scaling by 1e6 should work).

In more detail, when using a upper/lower bound, the internal value for the parameter is transformed to 
```python
    internal_value =  sqrt( (user_value +1 - min)**2 -1)
```
with ``min=0` and `user_value` 1.e-11, internal value is around 4.47e-6.   The fit then alters that internal value, first by taking a step that multiplies it by something like (1 + 1.5e-8).   That altered value is recast into user value to for your objective function.   

The problem is that recasting (4.5e-6)*(1+1.5e-8) back to the user value with 
```python
    user_value = min - 1 + sqrt(internal_value**2 + 1)
```
gives a value of  1.e-11, as the tiny change to the small value is lost in floating point math.  So, the objective function produces no change for the changed value and concludes the parameter has no effect on the objective function.

Generally speaking, leastsq (via MINPACK) does a very good job at automatically handling very different scales of parameter values, and one doesn't have to worry about parameters varying by 14 orders of magnitude.  Using bounds transformations as currently implemented in lmfit limits that scaling ability somewhat.

Hope that's clear.   We should probably include a discussion of this in the documentation.

**Comment by CarpeCimex** [comment]
@newville -- thanks for the careful explanation.  It seems like this is indeed what's happening.

Typically what I'm doing is fitting measured impedance data to models of electronic networks.  I will think about scaling my objective function.  There may be a way to scale all components, along with the impedance and frequency, and maintain the relationships between them.

I've been using leastsq directly for some time with these same small parameters, and you're right that it does a very good job with them.  I was hoping to add the boundary checking with lmfit and get the same good leastsq performance.  Have you considered implementing the bounds checking differently so it doesn't interfere with leastsq?  I was doing my own bounds checking by concatenating an array of numbers to my objective function result which got bigger the further out of bounds leastsq got, which worked okay, but I like the 'vary' and min/max limits of lmfit and would like to use them natively if possible.

**Comment by CarpeCimex** [comment]
@newville -- hey, I think I can make it work.  I found a frequency and impedance scaling method in an old filter design handbook, and it appears that if I scale my component values, frequency, and impedance the proper amounts before & after fitting I can avoid the limiting problem.

For the interested reader, here are the scaling factors to apply to all Rs, Cs, and Ls in a network and also the frequency and impedance:
    C' = C * FSF * ZSF
    L' = L
    R' = R / ZSF
    Z' = Z / ZSF
    w' = w / FSF
where
    FSF = Frequency Scaling Factor
    ZSF = Impedance (Z) Scaling Factor
Pick FSF and ZSF to bring your normalized (primed) component values into a good numerical range.  I used FSF = ZSF = 1e6 to convert pF -> units.  Convert all your denormalized (non-primed) values to normalized values as shown, use minimize() from lmfit on the normalized values, then denormalize everything again.  Pretty painless, and it seems to fit even better now than using leastsq directly as I was before, on the denormalized values.  Leastsq seems to like the normalized component values better.

**Comment by newville** [comment]
@CarpeCimex there is work from this summers GSOC that adds bounds to scipy.optimize.leastsq (see https://github.com/scipy/scipy/pull/5044) by adding a penalty when the Parameter is near the boundary,  and so is very different from the transformation used in lmfit (from leastsqbounds.py, from MINUIT).  My understanding is that this will be included in scipy 0.17.0.   Honestly, I've tried this exactly once.  I believe it will be better for your issue, and the benchmarks I saw implied better performance, but we'll have to test this out.

When it is readily available, we'll have to adapt lmfit leastsq() to be able to use this.  I haven't really thought about how to best do this.   An advantage of the current bounds transformations in lmfit is that it applies to the Parameters themselves, and so is not tied to a particular solver algorithm.  We'll for sure want to keep that, and may even want an option to use either bounds approach with lmfit leastsq().

Also: yeah, the fact that the normal engineering units for capacitance are Farads (instead of, say picoFarads) can really be a pain for numerical work.
